r/src/altrep.cpp

```diff
@@ -274,210 +257,342 @@ struct AltrepArrayPrimitive {
+      } else {
+        return Rf_ScalarInteger(static_cast<int>(value));
+      }
+    } else {
+      return Rf_ScalarReal(
+          internal::checked_cast<const DoubleScalar&>(*sum.scalar()).value);
+    }
   }
 };
 template <int sexp_type>
-R_altrep_class_t AltrepArrayPrimitive<sexp_type>::class_t;
+R_altrep_class_t AltrepVectorPrimitive<sexp_type>::class_t;
 
-// The methods below are how R interacts with the altrep objects.
-//
-// They all use the same pattern: create a C++ object of the
-// class parameter, and then call the method.
-template <typename AltrepClass>
-R_xlen_t Length(SEXP alt) {
-  return AltrepClass(alt).Length();
-}
+// Implementation for string arrays
+template <typename Type>
+struct AltrepVectorString : public AltrepVectorBase {
+  static R_altrep_class_t class_t;
+  using StringArrayType = typename TypeTraits<Type>::ArrayType;
 
-template <typename AltrepClass>
-Rboolean Inspect(SEXP alt, int pre, int deep, int pvec,
-                 void (*inspect_subtree)(SEXP, int, int, int)) {
-  return AltrepClass(alt).Inspect(pre, deep, pvec, inspect_subtree);
-}
+  static SEXP Make(const std::shared_ptr<Array>& array) {
+    return AltrepVectorBase::Make(class_t, array);
+  }
 
-template <typename AltrepClass>
-const void* Dataptr_or_null(SEXP alt) {
-  return AltrepClass(alt).Dataptr_or_null();
-}
+  // Get a single string, as a CHARSXP SEXP
+  // data2 is initialized, the CHARSXP is generated from the Array data
+  // and stored in data2, so that this only needs to expand a given string once
+  static SEXP Elt(SEXP alt_, R_xlen_t i) {
+    if (IsMaterialized(alt_)) {
+      return STRING_ELT(R_altrep_data2(alt_), i);
+    }
 
-template <typename AltrepClass>
-void* Dataptr(SEXP alt, Rboolean writeable) {
-  return AltrepClass(alt).Dataptr(writeable);
-}
+    // nul -> to NA_STRING
+    if (array(alt_)->IsNull(i)) {
+      return NA_STRING;
+    }
 
-template <typename AltrepClass>
-SEXP Duplicate(SEXP alt, Rboolean deep) {
-  return AltrepClass(alt).Duplicate(deep);
-}
+    // not nul, but we need care about embedded nuls
+    // this needs to call an R api function: Rf_mkCharLenCE() that
+    // might jump, i.e. throw an R error, which is dealt with using
+    // BEGIN_CPP11/END_CPP11/cpp11::unwind_protect()
+
+    BEGIN_CPP11
+
+    // C++ objects that will properly be destroyed by END_CPP11
+    // before it resumes the unwinding - and perhaps let
+    // the R error pass through
+    auto array_ = array(alt_);
+    auto view = internal::checked_cast<StringArrayType*>(array_.get())->GetView(i);
+    const bool strip_out_nuls = GetBoolOption("arrow.skip_nul", false);
+    bool nul_was_stripped = false;
+    std::string stripped_string;
+
+    // both cases might jump, although it's less likely when
+    // nuls are stripped, but still we need the unwind protection
+    // so that C++ objects here are correctly destructed, whilst errors
+    // properly pass through to the R side
```

Comment:
We need to handle both C++ destructors and R's jumping. `BEGIN_CPP11` / `END_CPP11` is essentially a C++ try/catch that knows how to deal with special exceptions `cpp11::unwind_exception`

These exceptions are thrown by `unwind_protect()` when the *C only* code it wraps jumps. When there is a jump, we have an opportunity to do any C++ cleanup (mostly destructors) before the unwinding is resumed so that the error is propagated back to the R side. 

In typical code that uses `cpp11` we don't see the `BEGIN_CPP11` / `END_CPP11` because they get generated by the `[[::export]]` decorations, but here we are not defining functions that we call from the R side, but rather defining functions that are called directly by R internals/altrep...  

https://github.com/r-lib/cpp11/blob/master/inst/include/cpp11/declarations.hpp

```cpp
#ifdef HAS_UNWIND_PROTECT
#define CPP11_UNWIND R_ContinueUnwind(err);
#else
#define CPP11_UNWIND \
  do {               \
  } while (false);
#endif

#define CPP11_ERROR_BUFSIZE 8192

#define BEGIN_CPP11                   \
  SEXP err = R_NilValue;              \
  char buf[CPP11_ERROR_BUFSIZE] = ""; \
  try {
#define END_CPP11                                               \
  }                                                             \
  catch (cpp11::unwind_exception & e) {                         \
    err = e.token;                                              \
  }                                                             \
  catch (std::exception & e) {                                  \
    strncpy(buf, e.what(), sizeof(buf) - 1);                    \
  }                                                             \
  catch (...) {                                                 \
    strncpy(buf, "C++ error (unknown cause)", sizeof(buf) - 1); \
  }                                                             \
  if (buf[0] != '\0') {                                         \
    Rf_errorcall(R_NilValue, "%s", buf);                        \
  } else if (err != R_NilValue) {                               \
    CPP11_UNWIND                                                \
  }                                                             \
  return R_NilValue;
```

Reply:
Hmm, thank you.